138945107 extract api service v5

tmikeschu opened the pull request:
Worked with @wlffann @kheppenstall

This PR extracts all API requests to a new model, PivotalTracker::ApiService

Some query methods are refactored so the ApiService methods can be more generalized.

Rebased and ready to rock. 🎸

tmikeschu committed:
Extract API request logic

Add PivotalTracker::ApiService model to handle API requests

[#138945107]

tmikeschu committed:
Extract API calls to new model

Add ::ApiService

Refactor project.stories to proxy and add #all to stories

[#138945107]

jasonnoble commented on lib/pivotal-tracker/account_membership.rb:
```diff
@@ -5,22 +5,12 @@ class PivotalTracker::AccountMembership < PivotalTracker::Resource
-    else		
-      raise(PivotalTracker::PermissionDenied, "Only Admins and Owners can see account memberships")		
-    end		
+    memberships = PivotalTracker::ApiService.all_nested('accounts', account_id, 'memberships')
```
raw_memberships?

jasonnoble replied:
Or `memberships_json`

tmikeschu replied:
👍

jasonnoble commented on lib/pivotal-tracker/api_service.rb:
```diff
@@ -0,0 +1,37 @@
+class PivotalTracker::ApiService
+  def self.find(collection, id)
```
Should this class be all the API related stuff currently at the top of pivotal_tracker.rb? i.e. HTTParty, base_uri, PermissionDenied error

tmikeschu replied:
👍

jasonnoble commented on lib/pivotal-tracker/api_service.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def self.all(collection)
+    PivotalTracker.get("/#{collection}").parsed_response
```
Add `if_success?` code here...

jasonnoble commented on lib/pivotal-tracker/api_service.rb:
```diff
@@ -0,0 +1,37 @@
+    if success?(response.code)
+      response.parsed_response
+    else
+      puts response.inspect
```
Hmmm.

jasonnoble commented on lib/pivotal-tracker/api_service.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def self.get_token(username, password)
+    PivotalTracker.basic_auth username, password
```
If we do https://github.com/jasonnoble/pivotal-tracker/pull/33/files#r99225728, this should just be `basic_auth username, password`.

jasonnoble commented on lib/pivotal-tracker/api_service.rb:
```diff
@@ -0,0 +1,37 @@
+
+  def self.get_token(username, password)
+    PivotalTracker.basic_auth username, password
+    PivotalTracker.get('/me')['api_token']
```
Remember, the `PivotalTracker` class shouldn't be doing `get` any more.  Should this be `me`?  i.e. `PivotalTracker::Api.me`?

jasonnoble commented on lib/pivotal-tracker/project.rb:
```diff
@@ -9,25 +9,16 @@ class PivotalTracker::Project < PivotalTracker::Resource
                 :velocity_averaged_over, :version, :week_start_day
   def self.all
-    PivotalTracker.get('/projects').parsed_response.map{|project| PivotalTracker::Project.new(project) }
+    projects = PivotalTracker::ApiService.all('projects')
```
`projects_json`?

tmikeschu replied:
👍

tmikeschu committed:
Address PR comments and refactor ApiService

Move parsed response logic to method
Raise one invalidity error
Change PivotalTracker to module

[#138945107]
